### PR TITLE
Filters: Short-circuit Variation calculations

### PR DESCRIPTION
**Pull request by johnrom**
Closed previous PRs due to [git flow branch naming requirements](https://github.com/woocommerce/woocommerce/wiki/WooCommerce-Git-Flow#branch-naming)

Relevant issue: #24659 (does not close)

Added short circuit filter for get_available_variations.
Added skip filter for Variation_Data_Store->read_price_data.

Test usage is being done on https://github.com/johnrom/woocommerce-variable-enhancements

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Added filters allowing third parties to short-circuit some very expensive variation calculations. Background tasks can then be used to generate this data over time for eventual consistency.

### How to test the changes in this Pull Request:

1. Generate thousands of variations on a product
2. Try to use the store or admin regarding this variable product, the site will not be responsive.
3. Add a filter of each type short circuiting expensive calculations.
4. Try to use the store or admin, the site should be responsive now, though there won't be "good" data.
5. Run a background task or something to generate said data in the background.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

I'm not certain new tests will apply here, since they would just test `add_filters()`.

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

```
Added short circuit filter for get_available_variations.
Added skip filter for Variation_Data_Store->read_price_data.
```

### CODE REVIEW

File: includes/data-stores/class-wc-product-variable-data-store-cpt.php [hunk 1]
```diff
@@ -286,84 +286,89 @@ public function read_price_data( &$product, $for_display = false ) {
...
 				);
 
-				$variation_ids = $product->get_visible_children();
+				// Allow 3rd parties to skip a full calculation of this data in favor of, for example, a background task.
+				$skip_price_data = apply_filters( 'woocommerce_variation_skip_price_data', false, $product );
 
-				if ( is_callable( '_prime_post_caches' ) ) {
-					_prime_post_caches( $variation_ids );
-				}
+				if ( ! $skip_price_data ) {
```
Comment (vedanshujain) [inline includes/data-stores/class-wc-product-variable-data-store-cpt.php]: can you move this condition as an `&&` clause inline 282? That way we don't have to indent the whole block. You might have to initialize `$transient_cached_prices_array[ $price_hash ]` though.

**Review by vedanshujain — COMMENTED** [review]
Yeah, filters as you suggest can be helpful. Left a minor comment though.

File: includes/data-stores/class-wc-product-variable-data-store-cpt.php (hunk 1 above)
Comment (johnrom) [reply includes/data-stores/class-wc-product-variable-data-store-cpt.php]: I was able to simplify the branching. Please have another look when you have a chance! Thanks.

**Comment by peterfabian** [comment]
Thanks for your contribution and apologies for the delay in the review.

While this change is quite small, well contained and allows for some easy performance wins, we are becoming a bit more cautious about adding filters per each function (and especially ones very much tied to a specific function, like `pre_get_available_variations`), as that quite considerably limits our opportunities for future improvements and refactorings. We'll discuss it a bit more internally and come back to you.

**Comment by claudiosanches** [comment]
Marked as "Blocked" since it requires some internal conversation.

**Comment by peterfabian** [comment]
Apologies that it took such a long time to come back to you and thanks for your willingness to improve WooCommerce! ❤️  

I think there are ultimately better ways to optimize this code, especially since short-circuiting those methods breaks assumptions other callees might have about these methods, so it seems using the filters can cause problems in the downstream code. It also goes against our principle to not use filters as feature flags for functions (https://github.com/woocommerce/woocommerce/wiki/Adding-Actions-and-Filters#using-filters-as-feature-flags). 

However, we understand this is an area that needs improvements, so we'd like to have a look at this more systematically at some point this year.

**Comment by johnrom** [comment]
Any ideas what other angles this can be attacked from? I'm going to be circling back to the project which will need optimization here in the coming months, and if there are any hints to which way we'll need to go, I can start a prototype as part of discovery for that project.

**Comment by peterfabian** [comment]
Just had a quick look now, I can see `get_children` is cached in a transient and all the variation objects would get cached within a request, too, via `_prime_post_caches` (or more persistently, if you have any of the caching plugins).

If you need a better result, I suggest trying to add some caching that would short circuit the `get_available_variations`. Theoretically, it needs to be invalidated on each stock level update, so the cache can be quite  short lived, but besides that, I'd assume most content won't be changing that often. Would that make sense to try this in your case?

**Pull request closed without merging**